### ISSUE TITLE
validation batch using more memory than training batch

### ISSUE THREAD

**Issue by Dankomaister** [atomistic-machine-learning/schnetpack#250]
Hi!
I was looking into the memory usage of schnetpack to find where it can be optimized and I found that the validation batches are using much more memory than the training batches.
Is this an expected behaviour? or is there something strange going on?

My tests use the same .db file for both training and validation (for the sake of testing) and on a Nvidia GTX 1070 with 8GB of VRAM I could use a maximum training batch size of 55 while the maximum validation batch size I could use was only 30 (while keeping the other batch size at 5, respectively).

I attached an MWE where there are 5 different sets of training batch, `T_batch_size`, and validation batch, `V_batch_size` sizes for testing.

/Daniel
[mem_test.zip](https://github.com/atomistic-machine-learning/schnetpack/files/5620169/mem_test.zip)

**Comment by Dankomaister** [comment]
Hi @mgastegger @Stefaanhess 
Any updates on this?

/Daniel

**Comment by Stefaanhess** [comment]
Hi @Dankomaister ,
sorry for the late reply!
I don´t think this is expected behaviour, but I will have a look at it today.
Stefaan

**Comment by Dankomaister** [comment]
Hi @Stefaanhess ,
Did you get anywhere on this?
I was thinking that maybe it is possible to set the `model.output_modules.create_graph = False` during validation?
Or do we need to create the graph during validation as well.
/Daniel

**Comment by Stefaanhess** [comment]
Hi @Dankomaister ,
sorry for the late reply!
I had a look at the train method and I think I found the issue.
During training and during validation, the computational graph is created. The difference between training and validation is, that the training loop calls `loss.backward()` for the optimization. This also frees the graph and reduces the memory allocation. Since the validation loop does not call `loss.backward()`, the graph is not freed when the next batch is forwarded through the model.
For models which are not trained on forces the validation loop could be computed with `torch.no_grad()`, but if forces are needed the graph is also needed.
Nevertheless I tried to detach the results (`val_result = {k: v.detach() for k, v in val_result.items()}`) and this reduces the allocated memory before the next batch is forwarded through the model. I created a new branch for this issue. What do you think about this?
Stefaan

**Comment by Dankomaister** [comment]
Hi @Stefaanhess ,
Detaching the val results sounds like a good solution, I will test and see how well it works.

The reason why I suggested to set `create_graph = False` during validation is that I have a script that I'm using to evaluate the predicted energies, forces and stress of a trained model against a .db file.
In this script I load a trained model using `model = load_model(model_file)` and evaluate it using the `evaluate_dataset` function from from `schnetpack.utils`.
What I found out is that if I set `model.output_modules[0].create_graph = False` after loading the model my memory usage during evaluation reduces from 7120MiB to 4856MiB for the same batch size and .db file.
The model is trained on energy, forces and stress and the predicted quantities are the same before and after setting  `model.output_modules[0].create_graph = False`

Correct me if I'm wrong but I thought that the computational graph is only required when performing backpropagation during training? and is not required when predicting?
Maybe it would be a good idea to implement a method to "freeze" a model after training, which optimizes the trained model for use only for predictions?
/Daniel

**Comment by Stefaanhess** [comment]
Hi @Dankomaister ,
you are right! ``create_graph=True`` is only needed for the backward pass. Maybe we can make use of the ``model.train()`` and ``model.eval()`` functions. In this case, the model will only create the graph in train mode. For the evaluation and the validation set, the model can be set to eval-mode. What do you think? I will talk to the others about this!
Stefaan

**Comment by Dankomaister** [comment]
Hi @Stefaanhess 
Ok that's good to know :)
I think it is a good idea to use `model.train()` and `model.eval()` to set the model to training or evaluation/validation mode.
Perhaps the `SpkCalculator` class could then automatically set the model to eval mode.
/Daniel

**Comment by Stefaanhess** [comment]
Hey @Dankomaister ,
I made a pull request for the _update_ branch, with the `train` and `eval` functions. `create_graph` is now turned off during the calculation of the validation set, during the evaluation of the model and for the `SpkCalculator`. What do you think? Does this solve the issue?
Stefaan
